My Site Dashboard [Phase 2]: Connect Dashboard Cards Store with Source

ParaskP7 opened the pull request:
Parent: #15498

This PR removes the mocked posts data and replaces that with the actual post data, which comes from the newly added `CardsStore` functionality. In effect, this PR connects the `CardsStore` with the `CardsSource` and makes sure that whenever this source is build, or refreshed for that matter, the data gets updated and shown on screen accordingly.

In more detail, this PR:
- Updates the naming convention and packaging structure on all things dashboard cards related (cards `source`, `update`, etc).
- Replaces the mocked posts data, and json utils, with the `CardsResult`, thus removing the need for the mocked posts data related json files altogether.
- Implements the new `CardsSource` logic, based on the `CardsStore`, getting the data from the database, through `Room`, utilizing flow, and fetching the data from the network, including posting the error state if necessary.

-----

To test:
- There is actually nothing to test in this PR, just review the `CardsSource` class and related changes and make sure that the accompanying tests are working as expected.
- When the [WordPress-FluxC-Android#2197](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2197) will get merged, and then this PR gets updated with the new version of FluxC, that which contains the update, then loading the dashboard feed for posts can be tested as a whole.

-----

To manual test:
- If you really want to test the solution, and I suggest you should:
  - Uncomment the `localFluxCPath = "../WordPress-FluxC-Android"` from your `local-builds.gradle` to work with you local `FluxC` repo.
  - Find the `issue/dashboard-cards-db` branch in `FluxC` and check it out.
  - Now build the project as usual and verify that
    - When the `MySiteDashboardPhase2FeatureConfig` is off, then the dashboard feed for posts doesn't appears on the `My Site` tab.
    - When the `MySiteDashboardPhase2FeatureConfig` is on, then the dashboard feed for posts does appears on the `My Site` tab, and with real, non-mocked data.
- However, mind the fact that you would also need to:
  - Either apply the [My Site Dashboard: Cards Endpoint](https://code.a8c.com/D69742) differential on your sandbox, proxy your device on it. Let me know if you need any help with that!
  - Or, you can wait for the [My Site Dashboard: Cards Endpoint](https://code.a8c.com/D69742) differential to be deployed. Then, you would be able to test the solution as usual.

-----

Merge instructions:
- [x] ~~Wait for the new API endpoint to be deployed (see [D69742](https://code.a8c.com/D69742)).~~
- [x] Wait for the accompanying [WordPress-FluxC-Android#2197](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2197) PR to be merged.
- [x] Update `fluxcVersion` with the once that includes the above FluxC solution (with either the `develop` hash or pointing to the newest release). PS: This PR will fail until this step is complete.
- [x] Remove the `[DO NOT MERGE]` from the title.
- [x] Remove the `[Status] Not Ready for Merge` label.
- [x] Merge this PR.

-----

## Regression Notes
1. Potential unintended areas of impact

When the `Phase 2` feature flag is:
- Off, then N/A.
- On, then the `My Site` tab and its inner cards and items might somehow be affected by the loading of the newly added `posts` dashboard cards.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

See `To test` and `To manual test` sections above.

3. What automated tests I added (or what prevented me from doing so)

N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ParaskP7 committed:
Refactor: Rename my site cards post package to dashboard

ParaskP7 committed:
Refactor: Rename post cards source to cards source

This source is actually going to be used for all dashboard related
cards, be it 'posts', 'stats', 'pages', 'comments', etc. As such, the
more specific 'post' naming is no longer appropriate and can be omitted.

Also, as with all the other 'cards' related naming convention, the
'Dashboard' word is omitted as well and used on the package level. Thus,
this makes the naming of the source as simple as possible, in this case
just 'CardsSource'. This also fits well with the 'CardsStore' on the
FluxC side.

ParaskP7 committed:
Refactor: Move post related cards to newly created posts package

This 'posts' package naming adheres to the 'posts' api endpoint
response property for all post card information related to posts, thus
this naming was chosen, instead the singular 'post' naming, to match the
pattern.

ParaskP7 committed:
Refactor: Optimize imports for cards builder test

ParaskP7 committed:
Refactor: Optimize imports for post card builder test

ParaskP7 committed:
Refactor: Rename posts update to cards update

This update is actually going to be used for all dashboard related
cards, be it 'posts', 'stats', 'pages', 'comments', etc. As such, the
more specific 'posts' naming is not appropriate and should be omitted.

ParaskP7 committed:
Refactor: Replace mocked posts data and json utils with cards result

Instead this commit replaces the 'MockedPostsData' with the actual
'CardsResult<List<CardModel>>' model, which the cards source is to
return.

At this moment, the 'CardsSource' return an empty 'CardsResult', but
subsequent commits will build on top of that to return either a non
empty or an error 'CardResult'.

Also, this commit updates the 'PostCardBuilder', which now uses the
actual data to build the individual post cards.

ParaskP7 committed:
Test: Update cards source test to adhere to cards store logic

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement initial cards store logic for cards source

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Feature: Add background dispatch to cards source

Since fetch cards is touching the network and database, this work needs
to be done on the background.

ParaskP7 committed:
Test: Add and update test for cards source based on get data and error

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement cards source logic based on get data and error

The way 'CardsSource' will provide the 'CardsUpdate' will be twofold:
1. First, the on 'build(...)' will get triggered. At that point the
'CardsStore' will hook on getting the cards from the database,
effectively collecting them from Room as a flow. At this point, any
available persisted/cached data will be displayed on screen immediately.
2. Then, and right after, a refresh of data will get triggered,
which will happen either immediately, through the initial 'build(...)'
trigger, or when 'refresh' get triggered on demand (for example due to
the user initiating a pull-to-refresh, or when the screen gets resumed).
At this point the 'CardsStore' will trigger a network request to fetch
the cards from the backend. When the network request returns, the
updated cards will be persisted in the database and the 'getData(...)'
hook will get automatically triggered (see 1. above).

In case of an error, during the fetching of the cards from the network,
the error state will be posted as well. Else, the success and empty
state of that network request will be ignored as it will be anyway
consumed by the flow, through Room (see 1. above).

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Cleanup: Remove mocked posts and refresh posts data json files

ParaskP7 committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into issue/15498-connect-dashboard-cards-store-with-source

 Conflicts:
	WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt
	WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt

ParaskP7 committed:
Merge: Update post cards source with card source on source manager

This update is necessary after merging this branch with 'develop' and
due to other merge conflicts due to the 'MySiteSourceManager' refactor
work.

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsSource.kt:
```diff
@@ -0,0 +1,82 @@
+        siteLocalId: Int
+    ) {
+        val selectedSite = selectedSiteRepository.getSelectedSite()
+        if (selectedSite != null && selectedSite.id == siteLocalId) {
```
@ParaskP7  At this point we could have flipped the refresh flag on, how are we flipping the flag off when the selectedSite.id != siteLocalId ?

ParaskP7 replied:
👋 @zwarm !

Good call, although this is just us guarding from an invalid state that we will probably never reach, you are right, we should be doing that, just like it happens on `ScanAndBackupSource`. I will update the solution to add a generic error when that happens, on both places, just like we do for the try/catch on `CardsStore`:

```
if (...) {
    ...
} else {
    postState(CardsUpdate(CardsResult(CardsError(GENERIC_ERROR))))
}
```

ParaskP7 replied:
Added here: a5d340b28feefa89209dc361c47cb49b84cc1dd5 + 83e0684c768ce5d151a858fd8e0232f4f738888d

zwarm replied:
🙇  It's so tricky to manage the refreshing indicator, thanks for adding this.

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsSource.kt:
```diff
@@ -0,0 +1,82 @@
+        coroutineScope.launch(bgDispatcher) {
+            val result = cardsStore.fetchCards(selectedSite)
+            result.error?.let {
+                postState(CardsUpdate(result))
```
Similar ❓  here ... if there is an error, we still need some type of update to flip the refresh flag to false? Thoughts?

ParaskP7 replied:
👋 @zwarm !

Thanks for the question. Actually, what you see there is that if there is an error, and only an error we are posting the state, thus you are seeing the `result.error?.let { ... }`. Then, the flow above will capture the rest.

Or, am I not understanding your question well?

zwarm replied:
Yes, I thought about this at 2AM ... this is only the error part. Thanks for clarifying.

zwarm commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsSourceTest.kt:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    @Test
+    fun `given no error, when data has been refreshed, then refresh is set to true`() = test {
```
Shouldn't refresh be set to false after a successful fetch? If not, the isRefreshing = true would result in the refreshing indicator never being hidden?

ParaskP7 replied:
👋 @zwarm !

Great question, but actually no, it is a bit tricky since I couldn't find a better naming and yet group both test somehow. So, what this test wants to verify is that during a refresh, when there is no error, then nothing happens to refresh, the flow above will capture the update, thus setting the refresh to false, not fetching the cards.

However, on the test further below, the one with an error given, then here we want to verify that the refresh is set to false. It is all about mixing and matching the `build(...)` + `getCards(...)` flow, and `refresh()` + `fetchCards(...)` result.

Can you suggest a better naming for both those tests so that their intention can be understood better?
- `given no error, when data has been refreshed, then refresh is set to true`
- `given error, when data has been refreshed, then refresh is set to false`

ParaskP7 replied:
Updated here: cf19cf099d297561ce5b364e2e8fb643e435bc5c

Maybe the update I did here will help you understand what this test is doing and why. 🤔

zwarm replied:
@ParaskP7 Yes, that comment helped. 

What tripped me up when I read this is the past tense in the when portion` **has been refreshed**`. I equate that with the process is complete and the final value should be set, but in this case the flag flip happens elsewhere, so the value will always be true. The comments on the verification are helpful.

ParaskP7 replied:
I am glad it helped @zwarm but all you said are right, I just had hard time thinking of a better alternative for naming things. Feel free to suggest anything else and I can incorporate it on a subsequent commit without problem. 🙏

zwarm reviewed (changes requested):
👋  Nice refactoring on the names 🙇 . Much easier to follow and sets us up for the future.

I left two questions on `CardsSource` and one on `CardSourceTest` about refreshing. This very well may be the same issue we already set aside for brainstorming.

I have not done any end-to-end testing, I've only run the tests in this PR and they work as expected. I will hold off on approval until after I can test end-to-end. Plus it will give you some time to answer my questions. Thanks 🔆

ParaskP7 committed:
Test: Add tests for invalid site on cards source

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement invalid site logic on cards source

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Test: Update cards source tests by observing refresh before build

By doing that, the data refreshed tests become more deterministic and
easier to follow. Also, as part of this update, and for those data
refreshed tests, extra comments were added to explain what is asserted
and why.

ParaskP7 commented:
👋 @zwarm !

> 👋 Nice refactoring on the names 🙇 . Much easier to follow and sets us up for the future.

Yea, I believe this will help us going forward, I am glad you liked it! 💪 

> I left two questions on CardsSource and one on CardSourceTest about refreshing. This very well may be the same issue we already set aside for brainstorming.

Thanks for all your questions. I actually pushed few commits to address some of them! 🥇 

> I have not done any end-to-end testing, I've only run the tests in this PR and they work as expected. I will hold off on approval until after I can test end-to-end. Plus it will give you some time to answer my questions. Thanks 🔆

Yes, 💯  and thank you! 🙏

zwarm reviewed (approved):
Sweet success. @ParaskP7 - everything looks fantastic. Nice work!! 🚀

zwarm committed:
Update the FluxC hash

zwarm committed:
Merge branch 'develop' into issue/15498-connect-dashboard-cards-store-with-source
